src/joystick/linux/SDL_sysjoystick.c

```diff
@@ -260,18 +260,20 @@ static int IsJoystick(const char *path, int fd, char **name_return, SDL_Joystick
     struct input_id inpid;
     char *name;
     char product_string[128];
+    SDL_UDEV_deviceclass class = 0;
 
-    if (ioctl(fd, JSIOCGNAME(sizeof(product_string)), product_string) >= 0) {
-        SDL_zero(inpid);
+    SDL_zero(inpid);
 #ifdef SDL_USE_LIBUDEV
-        SDL_UDEV_GetProductInfo(path, &inpid.vendor, &inpid.product, &inpid.version);
+    SDL_UDEV_GetProductInfo(path, &inpid.vendor, &inpid.product, &inpid.version, &class);
 #endif
-    } else {
+    if (ioctl(fd, JSIOCGNAME(sizeof(product_string)), product_string) <= 0) {
         /* When udev is enabled we only get joystick devices here, so there's no need to test them */
-        if (enumeration_method != ENUMERATION_LIBUDEV && !GuessIsJoystick(fd)) {
+        if ( enumeration_method != ENUMERATION_LIBUDEV &&
+             !(class & SDL_UDEV_DEVICE_JOYSTICK) && !GuessIsJoystick(fd) ) {
```

Comment:
I'm not sure about this logic. You've retrieved the udev code path's best guess at the device class, and you are using it for 1 bit of information: "is it believed to be a joystick?"

It might make more sense to treat the device class bitmask as a tristate:

1. believed to be a joystick (`class & SDL_UDEV_DEVICE_JOYSTICK`)
2. believed to be something other than a joystick (`class != 0 && !(class & SDL_UDEV_DEVICE_JOYSTICK)`)
3. no idea (`class == 0`)

and only go into `GuessIsJoystick` if we're in the third case?

Or maybe `class` should be a parameter to `GuessIsJoystick`, so that it can override our guesses from the evdev capabilities.

Reply:
I implemented the tri-state logic with
```
        if ( enumeration_method != ENUMERATION_LIBUDEV &&
             !(class & SDL_UDEV_DEVICE_JOYSTICK) && ( class || !GuessIsJoystick(fd) ) ) {
            return 0;
        }
```
That is, return if either
* not udev enumeration, not udev joystick class, but some (other) udev class, or
* not udev enumeration, not udev joystick class, not some (other) udev class, and not guessed to be a joystick

or, from the other perspective, only continue if
* udev enumeration,
* udev joystick class, or
* not some (other) udev class and guessed to be a joystick